### PR TITLE
Synchronizing populatePropertyBag as additional fallback to avoid tra…

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**
…nsient invalid json in Changefeed state

It is not fully clear yet how we occasionally end-up with invalid json for the ChangeFeedStartFrom settings (json would have "Type": "Now", "Type": "Now") - the populate property bag method uses Map<>.put (LinkedHashMap) to add the property - which should never result in duplicates. Synchronizing this part is just an attempt to reduce risk - not because there is a known multi-threaded usage that could result in the duplicates.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromBeginningImpl.java [hunk 1]
```diff
@@ -16,10 +16,12 @@ public ChangeFeedStartFromBeginningImpl() {
     public void populatePropertyBag() {
         super.populatePropertyBag();
 
-        setProperty(
-            this,
-            Constants.Properties.CHANGE_FEED_START_FROM_TYPE,
-            ChangeFeedStartFromTypes.BEGINNING);
+        synchronized(this) {
```
Comment (milismsft) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromBeginningImpl.java:19]: Does base class method call need to be "synchronized"/single threaded as well?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromETagAndFeedRangeImpl.java [hunk 2]
```diff
@@ -82,19 +82,21 @@ public void populateRequest(RxDocumentServiceRequest request) {
...
-
-        setProperty(
-            this,
-            Constants.Properties.CHANGE_FEED_START_FROM_ETAG,
-            this.eTag);
-
-        this.feedRange.setProperties(
-            this,
-            true);
+        synchronized(this) {
```
Comment (milismsft) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromETagAndFeedRangeImpl.java:85]: Does base class method call need to be "synchronized"/single threaded as well?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromNowImpl.java [hunk 3]
```diff
@@ -16,12 +16,15 @@ public ChangeFeedStartFromNowImpl() {
...
     @Override
     public void populatePropertyBag() {
-        super.populatePropertyBag();
-
-        setProperty(
-            this,
-            Constants.Properties.CHANGE_FEED_START_FROM_TYPE,
-            ChangeFeedStartFromTypes.NOW);
+        
+            super.populatePropertyBag();
```
Comment (milismsft) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromNowImpl.java:20]: Does base class method call need to be "synchronized"/single threaded as well?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromPointInTimeImpl.java [hunk 4]
```diff
@@ -49,15 +49,17 @@ public void populatePropertyBag() {
...
-        setProperty(
-            this,
-            com.azure.cosmos.implementation.Constants.Properties.CHANGE_FEED_START_FROM_TYPE,
-            ChangeFeedStartFromTypes.POINT_IN_TIME);
-
-        setProperty(
-            this,
-            Constants.Properties.CHANGE_FEED_START_FROM_POINT_IN_TIME_MS,
-            this.pointInTime.toEpochMilli());
+        synchronized(this) {
```
Comment (milismsft) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromPointInTimeImpl.java:52]: Does base class method call need to be "synchronized"/single threaded as well?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromBeginningImpl.java [hunk 5]
```diff
@@ -16,10 +16,12 @@ public ChangeFeedStartFromBeginningImpl() {
     public void populatePropertyBag() {
         super.populatePropertyBag();
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromBeginningImpl.java:17]: if there is any concurrency concern. I suspect the synchronization may need to get extend here as well.

any reason this needs to be outside of synchronization scope.

also elsewhere for similar cases.

**Review by moderakh — APPROVED** [review]
LGTM. one question on why `super.populatePropertyBag(); doesn't need to get included in the synchronization scope?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromBeginningImpl.java (hunk 5 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromBeginningImpl.java:17]: The base class' ChangeFeedStartFromInternal.populatePropertyBag is a no-op as is its baseclass JsonSerializable's populatePropertyBag method.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromETagAndFeedRangeImpl.java (hunk 2 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromETagAndFeedRangeImpl.java:85]: The base class' ChangeFeedStartFromInternal.populatePropertyBag is a no-op as is its baseclass JsonSerializable's populatePropertyBag method.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromNowImpl.java (hunk 3 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromNowImpl.java:20]: The base class' ChangeFeedStartFromInternal.populatePropertyBag is a no-op as is its baseclass JsonSerializable's populatePropertyBag method.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromPointInTimeImpl.java (hunk 4 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromPointInTimeImpl.java:52]: The base class' ChangeFeedStartFromInternal.populatePropertyBag is a no-op as is its baseclass JsonSerializable's populatePropertyBag method.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromBeginningImpl.java (hunk 1 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromBeginningImpl.java:19]: The base class' ChangeFeedStartFromInternal.populatePropertyBag is a no-op as is its baseclass JsonSerializable's populatePropertyBag method.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromETagAndFeedRangeImpl.java (hunk 2 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromETagAndFeedRangeImpl.java:85]: The base class' ChangeFeedStartFromInternal.populatePropertyBag is a no-op as is its baseclass JsonSerializable's populatePropertyBag method.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromNowImpl.java (hunk 3 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromNowImpl.java:20]: The base class' ChangeFeedStartFromInternal.populatePropertyBag is a no-op as is its baseclass JsonSerializable's populatePropertyBag method.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromPointInTimeImpl.java (hunk 4 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromPointInTimeImpl.java:52]: The base class' ChangeFeedStartFromInternal.populatePropertyBag is a no-op as is its baseclass JsonSerializable's populatePropertyBag method.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromBeginningImpl.java (hunk 5 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/ChangeFeedStartFromBeginningImpl.java:17]: The base class' ChangeFeedStartFromInternal.populatePropertyBag is a no-op as is its baseclass JsonSerializable's populatePropertyBag method.

**Pull request merged**